packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,75 @@ class MaterialAccentColor extends ColorSwatch<int> {
   Color get shade700 => this[700];
 }
 
+/// Signature for the function that returns a (potentially different) color
+/// based on a given set of states.
+typedef ColorFunction = Color Function(Set<MaterialState> states);
+
+/// Defines a [Color] that can react to changes in the state of a Material
+/// component, based on a given set of [MaterialState]s.
+///
+/// This is useful for preserving the accessibility of text in different states
+/// of a component. For example, in a [TextButton] with blue text, the text will
+/// become more difficult to read when the button is hovered, focused, or pressed,
+/// because the contrast ratio between the button and the text will dip. To solve
+/// this, you can use [MaterialStateColor] to make the text darker when the
+/// [TextButton] is hovered, focused, or pressed.
+///
+/// The [colorFunction] member is the callback that will be used to get the color
+/// in a given context.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
+/// ```dart
+/// Color getTextColor(Set<MaterialSet> states) {
+///   final interactiveStates = <MaterialState>{
+///     MaterialState.pressed,
+///     MaterialState.hovered,
+///     MaterialState.focused,
+///   };
+///   if(states.any(interactiveStates.contains) {
+///     return Colors.blue[900];
+///   }
+///   return Colors.blue[600];
+/// }
+///
+/// FlatButton(
+///   ...
+///   textColor: MaterialStateColor(getTextColor),
+/// ),
+/// ```
+class MaterialStateColor extends Color {
+  /// Creates a [MaterialStateColor], a color that can return different values
+  /// depending on a given [MaterialState].
+  ///
+  /// The callback must return a non-null color in the default state (empty set).
+  MaterialStateColor(this.colorFunction) :
+    assert(colorFunction(_defaultState) != null),
+    super(colorFunction(_defaultState).value);
+
+  /// The callback that returns a [Color] for a given state.
+  ///
+  /// The callback must return a non-null color in the default state (empty set).
+  final ColorFunction colorFunction;
+
+  /// The default state for a Material component, the empty set of interaction states.
+  static const Set<MaterialState> _defaultState = <MaterialState>{};
+
+  /// Returns the color for the given set of states if `color` is a
```

Comment:
Havn't seen this in context yet, but when color is a Color, it does not need states, but it still has to be passed. Is it ever unknown what instance Color is? Can it be checked at that time before calling this method?

Reply:
This is a convenience method so that the button class doesn't have to do `color is MaterialStateColor ? color.resolve(states) : color;`. You would only ever call this method if you had states you wanted to pass to it, so it makes sense that states is required.